db/upgrade.php

```diff
@@ -126,5 +126,54 @@ function xmldb_reengagement_upgrade($oldversion=0) {
         upgrade_mod_savepoint(true, 2017040400, 'reengagement');
     }
 
+    if ($oldversion < 2017102001) {
+        // A bug in previous versions prevented some e-mails from being sent.
+        // Flag these old broken reengagment e-mails as sent as they may no longer be relevant.
+        $timenow = time();
+        $sql = "SELECT rin.*, cm.id as cmid, r.emailuser from mdl_reengagement_inprogress rin
+                  JOIN {course_modules} cm ON cm.instance = rin.reengagement
+                  JOIN {modules} m ON m.id = cm.module
+                  JOIN {reengagement} r on r.id = rin.reengagement
+             LEFT JOIN {course_modules_completion} cmc ON cmc.coursemoduleid = cm.id AND cmc.userid = rin.userid
+                 WHERE m.name = 'reengagement' AND cmc.id is null
+                       AND rin.completiontime < ? AND rin.completed = 0";
+        $missingprogress = $DB->get_recordset_sql($sql, array($timenow));
+
+        foreach ($missingprogress as $missing) {
+            // Add course completion record.
+            $activitycompletion = new stdClass();
+            $activitycompletion->coursemoduleid = $missing->cmid;
+            $activitycompletion->completionstate = COMPLETION_COMPLETE_PASS;
+            $activitycompletion->timemodified = $timenow;
+            $activitycompletion->userid = $missing->userid;
+            $DB->insert_record('course_modules_completion', $activitycompletion);
+
+            // flag re-enagement as complete if required - or delete record.
+            // logic copied from cron function.
+            if (($missing->emailuser == REENGAGEMENT_EMAILUSER_COMPLETION) ||
+                ($missing->emailuser == REENGAGEMENT_EMAILUSER_NEVER) ||
+                ($missing->emailuser == REENGAGEMENT_EMAILUSER_TIME && !empty($missing->emailsent))) {
+                // No need to keep 'inprogress' record for later emailing
+                // Delete inprogress record.
+                debugging('', DEBUG_DEVELOPER) && mtrace("mode $missing->emailuser reengagementid $missing->id.
+                      User marked complete, deleting inprogress record for user $missing->userid");
+                $DB->delete_records('reengagement_inprogress', array('id' => $missing->id));
+            } else {
+                // Update inprogress record to indicate completion done.
+                debugging('', DEBUG_DEVELOPER) && mtrace("mode $missing->emailuser reengagementid $missing->id
+                      updating inprogress record for user $missing->userid to indicate completion");
+                $updaterecord = new stdClass();
+                $updaterecord->id = $missing->id;
+                $updaterecord->completed = COMPLETION_COMPLETE;
```

Comment:
Not defined - could include mod/reengagement/lib.php which includes completionlib.php